### PR TITLE
Rewriting URLs for images hosted on Google to higher resolution versions #3894

### LINKED ISSUE #3894: Rewriting URLs for images hosted on Google to higher resolution versions
Related to: #3860

Some of the NFT collection images aren't sharp. Eg. Area's:

Screenshot 2022-02-10 at 11 05 37 AM

It seems like many of the images in the OpenSea API output are hosted on Google:

```
"image_url" : "https://lh3.googleusercontent.com/P1bYlMN-oZ5iqoq2bvsAwFGkZaZWhxE4baRzaLVX2NjEE57l1otA23aE8GChhdV5OL8tMdQbeynbllMw5hMQX8cKVpL0j_nYKE74PA=s120",
"banner_image_url" : "https://lh3.googleusercontent.com/FUDEOcpQOFILbySm_Kd0DyTtxdubU-pUy4WGxkipcmOWXLBCbMUSJ78-tSTKd9g_-NGAso3SZSZjL-MWHmwT2JIUTecNWcLlnHoRRw=s2500",
"image_url" : "https://lh3.googleusercontent.com/P1bYlMN-oZ5iqoq2bvsAwFGkZaZWhxE4baRzaLVX2NjEE57l1otA23aE8GChhdV5OL8tMdQbeynbllMw5hMQX8cKVpL0j_nYKE74PA=s120",
"image_preview_url" : "https://lh3.googleusercontent.com/Kr8PimjwslvBMGk-JhAvIirmlAQ4wLuZ-U1mWsCB27cg2XjKxj7WEis4TPJg440A5kYREQ-z6jtMIIresbEEJH--jbRb9P6p6o6D-w=s250",
"image_thumbnail_url" : "https://lh3.googleusercontent.com/Kr8PimjwslvBMGk-JhAvIirmlAQ4wLuZ-U1mWsCB27cg2XjKxj7WEis4TPJg440A5kYREQ-z6jtMIIresbEEJH--jbRb9P6p6o6D-w=s128",
"featured_image_url" : "https://lh3.googleusercontent.com/41b_hLdXm_qEg7g-6eVTeeU8Q_hw8irj-PExebDc4zgtHjyUPV8tRY6mebfaz_vCJOcMnZ0CRKGo4U_bR77t83PvbWf020XBnw7YPfM=s300",
```

According to this reference (not sure if it's correct, but seems to work anecdotally): https://developers.google.com/people/image-sizing

We can convert from the original to a specific width/height (the longer side) like this, from:

`https://lh3.googleusercontent.com/P1bYlMN-oZ5iqoq2bvsAwFGkZaZWhxE4baRzaLVX2NjEE57l1otA23aE8GChhdV5OL8tMdQbeynbllMw5hMQX8cKVpL0j_nYKE74PA=s120`

to:

`https://lh3.googleusercontent.com/P1bYlMN-oZ5iqoq2bvsAwFGkZaZWhxE4baRzaLVX2NjEE57l1otA23aE8GChhdV5OL8tMdQbeynbllMw5hMQX8cKVpL0j_nYKE74PA=s750` (s120 -> s750)

with whichever dimension is appropriate, ending up with:

from OpenSea|after rewriting URL
-|-
 | 

The [-p (smart crop is also interesting)](https://lh3.googleusercontent.com/P1bYlMN-oZ5iqoq2bvsAwFGkZaZWhxE4baRzaLVX2NjEE57l1otA23aE8GChhdV5OL8tMdQbeynbllMw5hMQX8cKVpL0j_nYKE74PA=s750-p) but maybe only obvious if we have non-square images:

But defaulting to smart crop might be useful, performance-wise?

Do you think this is worth doing and not fragile? If we do it, let's create a new type like `AlphaWallet.Address` which wraps image URLs.

### PR DESCRIPTION
**Pull request by oa-s**
Closes #3894

### CODE REVIEW

**Comment by hboon** [comment]
@oa-s will take a look. But some quick feedback:

* Copyright says 2020
* The README mentions your gmail address (if you are ok with it)

File: modules/AlphaWalletCore/AlphaWalletCore/WebImageURL.swift [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+//  AlphaWalletCore
+//
+//  Created by Vladyslav Shepitko on 11.02.2022.
+//
+
+import Foundation
+
+public enum WebImageURL: Codable, Hashable, Equatable, CustomStringConvertible {
+    case origin(URL)
+    case googleContentRewritten(URL)
```
Comment (hboon) [inline modules/AlphaWalletCore/AlphaWalletCore/WebImageURL.swift:12]: Are we rewriting the URL for Google hosted images for all cases? Should we add a case or somehow be able to choose when to use the original. Eg. we only want to use the rewritten, bigger version when in Wallet > Collectibles, and Info, and in the screen showing a single token ID, but not in screens like the little icons in Wallet tab or Activity Tab. 

Concerned with memory usage and speed.

**Review by hboon — CHANGES_REQUESTED** [review]
Looks great so far! Thanks!

File: modules/AlphaWalletCore/AlphaWalletCore/WebImageURL.swift [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+    public var absoluteString: String {
+        return url.absoluteString
+    }
+
+    public var description: String {
+        return absoluteString
+    }
+
+    public var url: URL {
```
[the remark this answers is not in the document]
Comment (oa-s) [reply modules/AlphaWalletCore/AlphaWalletCore/WebImageURL.swift:23]: made it as public as we need back convertation to `URL`

**Comment by oa-s** [comment]
@hboon Its ready.

**Comment by hboon** [comment]
Merge when you are ready

**Pull request merged**